<a href="https://colab.research.google.com/github/aimedvedeva/Competition1/blob/main/competition_1_preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>





# Reading  Data from Github

In [61]:
import pandas as pd
import io
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px

In [40]:
import requests
import pandas as pd

def get_github_file_contents(repo_owner, repo_name, folder_path, file_name):
    raw_url = f'https://raw.githubusercontent.com/{repo_owner}/{repo_name}/master/{folder_path}/{file_name}'
    response = requests.get(raw_url)
    if response.status_code == 200:
        return response.text
    else:
        print(f"Error: Unable to fetch data")
        return None


In [47]:
repo_owner = 'aimedvedeva'
repo_name = 'Competition1'
folder_path = 'data'
train = pd.read_csv(io.StringIO(get_github_file_contents(repo_owner, repo_name, folder_path, 'train.csv')))
test = pd.read_csv(io.StringIO(get_github_file_contents(repo_owner, repo_name, folder_path, 'test.csv')))
meta_data = pd.read_csv(io.StringIO(get_github_file_contents(repo_owner, repo_name, folder_path, 'greeks.csv')))


In [46]:
train.head()

Id        AB          AF          AH         AM        AR  \
0  000ff2bfdfe9  0.209377  3109.03329   85.200147  22.394407  8.138688   
1  007255e47698  0.145282   978.76416   85.200147  36.968889  8.138688   
2  013f2bd269f5  0.470030  2635.10654   85.200147  32.360553  8.138688   
3  043ac50845d5  0.252107  3819.65177  120.201618  77.112203  8.138688   
4  044fb8a146ec  0.380297  3733.04844   85.200147  14.103738  8.138688   

         AX        AY         AZ          BC  ...        FL        FR  \
0  0.699861  0.025578   9.812214    5.555634  ...  7.298162   1.73855   
1  3.632190  0.025578  13.517790    1.229900  ...  0.173229   0.49706   
2  6.732840  0.025578  12.824570    1.229900  ...  7.709560   0.97556   
3  3.685344  0.025578  11.053708    1.229900  ...  6.122162   0.49706   
4  3.942255  0.054810   3.396778  102.151980  ...  8.153058  48.50134   

         FS         GB          GE            GF         GH         GI  \
0  0.094822  11.339138   72.611063   2003.810319  22.136229  69.834944   
1  0.568932   9.292698   72.611063  27981.562750  29.135430  32.131996   
2  1.198821  37.077772   88.609437  13676.957810  28.022851  35.192676   
3  0.284466  18.529584   82.416803   2094.262452  39.948656  90.493248   
4  0.121914  16.408728  146.109943   8524.370502  45.381316  36.262628   

          GL  Class  
0   0.120343      1  
1  21.978000      0  
2   0.196941      0  
3   0.155829      0  
4   0.096614      1  

[5 rows x 58 columns]

In [48]:
test.head()

Id   AB   AF   AH   AM   AR   AX   AY   AZ   BC  ...   FI   FL  \
0  00eed32682bb  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1  010ebe33f668  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
2  02fa521e1838  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
3  040e15f562a2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
4  046e85c7cc7f  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   

    FR   FS   GB   GE   GF   GH   GI   GL  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 57 columns]

In [50]:
meta_data.head()

Id Alpha Beta Gamma Delta    Epsilon
0  000ff2bfdfe9     B    C     G     D  3/19/2019
1  007255e47698     A    C     M     B    Unknown
2  013f2bd269f5     A    C     M     B    Unknown
3  043ac50845d5     A    C     M     B    Unknown
4  044fb8a146ec     D    B     F     B  3/25/2020

# EDA

In [58]:
def EDA(df):
    print(f'data shape: {df.shape}')
    res = pd.DataFrame(df.dtypes, columns=['data type'])
    res['#missing'] = df.isnull().sum().values
    res['%missing'] = df.isnull().sum().values / len(df)* 100
    res['#unique'] = df.nunique().values
    desc = pd.DataFrame(df.describe(include='all').transpose())
    res['min'] = desc['min'].values
    res['max'] = desc['max'].values
    res['first quartile'] = desc.loc[:, '25%'].values
    res['second quartile'] = desc.loc[:, '50%'].values
    res['third quartile'] = desc.loc[:, '75%'].values
    print(f"total missing data : {sum(res['#missing'])}")

    return res

In [59]:
EDA(train)

data shape: (617, 58)
total missing data : 131


data type  #missing  %missing  #unique          min          max  \
Id       object         0  0.000000      617          NaN          NaN   
AB      float64         0  0.000000      217     0.081187     6.161666   
AF      float64         0  0.000000      599    192.59328  28688.18766   
AH      float64         0  0.000000      227    85.200147  1910.123198   
AM      float64         0  0.000000      605     3.177522    630.51823   
AR      float64         0  0.000000      130     8.138688   178.943634   
AX      float64         0  0.000000      427     0.699861     38.27088   
AY      float64         0  0.000000      148     0.025578    10.315851   
AZ      float64         0  0.000000      484     3.396778    38.971568   
BC      float64         0  0.000000      259       1.2299  1463.693448   
BD      float64         0  0.000000      617   1693.62432  53060.59924   
BN      float64         0  0.000000       53       9.8868      29.3073   
BP      float64         0  0.000000      612    72.948951   2447.81055   
BQ      float64        60  9.724473      515     1.331155   344.644105   
BR      float64         0  0.000000      566    51.216883  179250.2529   
BZ      float64         0  0.000000      115   257.432377   50092.4593   
CB      float64         2  0.324149      553     12.49976  2271.436167   
CC      float64         3  0.486224      602     0.176874     4.103032   
CD      float64         0  0.000000      584      23.3876   633.534408   
CF      float64         0  0.000000      586     0.510888   200.967526   
CH      float64         0  0.000000      135     0.003184     0.224074   
CL      float64         0  0.000000      123     1.050225    31.688153   
CR      float64         0  0.000000      595     0.069225     3.039675   
CS      float64         0  0.000000      576    13.784111   267.942823   
CU      float64         0  0.000000      307     0.137925     4.951507   
CW      float64         0  0.000000      426      7.03064    64.521624   
DA      float64         0  0.000000      611       6.9064    210.33092   
DE      float64         0  0.000000      616    35.998895   2103.40519   
DF      float64         0  0.000000      137      0.23868    37.895013   
DH      float64         0  0.000000      191     0.040995     1.060404   
DI      float64         0  0.000000      571     60.23247  1049.168078   
DL      float64         0  0.000000      604      10.3456     326.2362   
DN      float64         0  0.000000      576     6.339496    62.808096   
DU      float64         1  0.162075      253     0.005518   161.355315   
DV      float64         0  0.000000       39      1.74307     25.19293   
DY      float64         0  0.000000      590     0.804068   152.355164   
EB      float64         0  0.000000      439     4.926396     94.95858   
EE      float64         0  0.000000      513     0.286201    18.324926   
EG      float64         0  0.000000      610     185.5941  30243.75878   
EH      float64         0  0.000000      127     0.003042    42.569748   
EJ       object         0  0.000000        2          NaN          NaN   
EL      float64        60  9.724473      311     5.394675   109.125159   
EP      float64         0  0.000000      275    78.526968  1063.594578   
EU      float64         0  0.000000      455     3.828384   6501.26448   
FC      float64         1  0.162075      600     7.534128  3030.655824   
FD      float64         0  0.000000      337      0.29685  1578.654237   
FE      float64         0  0.000000      615  1563.136688  143224.6823   
FI      float64         0  0.000000      498      3.58345    35.851039   
FL      float64         1  0.162075      388     0.173229   137.932739   
FR      float64         0  0.000000      435      0.49706   1244.22702   
FS      float64         2  0.324149      161      0.06773    31.365763   
GB      float64         0  0.000000      560     4.102182   135.781294   
GE      float64         0  0.000000      264    72.611063  1497.351958   
GF      fl

**Observations**

There are 57 variables (X) and 1 variable (y)

tota number if missing data is 131 and mostly for varalbes "BQ" and "EL"

In the first step we ignore the missing values as the total number is quite low.

All variables are numeric except "ID" and "EJ"

**Checking the dataset balance**

In [63]:
fig2 = px.pie(train, names='Class',
             height=400, width=600,
             hole=0.7,
             title='target class Overview',
                   color_discrete_sequence=['#4c78a8', '#72b7b2'])
fig2.update_traces(hovertemplate=None, textposition='outside', textinfo='percent+label', rotation=0)
fig2.update_layout(margin=dict(t=100, b=30, l=0, r=0), showlegend=False,
                        plot_bgcolor='#fafafa', paper_bgcolor='#fafafa',
                        title_font=dict(size=20, color='#555', family="Lato, sans-serif"),
                        font=dict(size=17, color='#8a8d93'),
                        hoverlabel=dict(bgcolor="#444", font_size=13, font_family="Lato, sans-serif"))
fig2.show()

**Observation**

The target variable is imbalanced

We can use some methods like RandomForest which handles imbalanced data or we can use undersampling